src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,566 @@
+ *     to not have an end-of-stream hook and invoking their finisher is
+ *     optional.</li>
+ * </ul>
+ *
+ * @see Stream#gather(Gatherer)
+ * @see Gatherers
+ *
+ * @param <T> the type of input elements to the gatherer operation
+ * @param <A> the potentially mutable state type of the gatherer operation
+ *            (often hidden as an implementation detail)
+ * @param <R> the type of output elements from the gatherer operation
+ * @since 22
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+public interface Gatherer<T, A, R> {
+    /**
+     * A function that produces an instance of the intermediate state used for
+     * this gathering operation.
+     *
+     * <p>By default, this method returns {@link #defaultInitializer()}
+     *
+     * @return A function that produces an instance of the intermediate state
+     * used for this gathering operation
+     */
+    default Supplier<A> initializer() {
+        return defaultInitializer();
+    };
+
+    /**
+     * A function which integrates provided elements, potentially using
+     * the provided intermediate state, optionally producing output to the
+     * provided {@link Downstream}.
+     *
+     * @return a function which integrates provided elements, potentially using
+     *         the provided state, optionally producing output to the provided
+     *         Downstream
+     */
+    Integrator<A, T, R> integrator();
+
+    /**
+     * A function which accepts two intermediate states and combines them into
+     * one.
+     *
+     * <p>By default, this method returns {@link #defaultCombiner()}
+     *
+     * @return a function which accepts two intermediate states and combines
+     *         them into one
+     */
+    default BinaryOperator<A> combiner() {
+        return defaultCombiner();
+    }
+
+    /**
+     * A function which accepts the final intermediate state
+     * and a {@link Downstream} object, allowing to perform a final action at
+     * the end of input elements.
+     *
+     * <p>By default, this method returns {@link #defaultFinisher()}
+     *
+     * @return a function which transforms the intermediate result to the final
+     *         result(s) which are then passed on to the provided Downstream
+     */
+    default BiConsumer<A, Downstream<? super R>> finisher() {
+        return defaultFinisher();
+    }
+
+    /**
+     * Returns a composed Gatherer which connects the output of this Gatherer
+     * to the input of that Gatherer.
+     *
+     * @param that the other gatherer
+     * @param <AA> The type of the state of that Gatherer
+     * @param <RR> The type of output of that Gatherer
+     * @throws NullPointerException if the argument is null
+     * @return returns a composed Gatherer which connects the output of this
+     *         Gatherer as input that Gatherer
+     */
+    default <AA, RR> Gatherer<T, ?, RR> andThen(Gatherer<? super R, AA, ? extends RR> that) {
```

Comment:
It would be great to remove AA type parameter, which is mentioned only once, so could be replaced with `?`. This will improve user experience if they happen to need to specify type parameters explicitly.

Reply:
This is especially important given that often you don't know the AA type at all. E.g., imagine that you are doing `.andThen(Gatherers.fold(...))`, but `fold` returns `Gatherer<T, ?, R>`, so you can specify explicit `RR`, but not `AA`.

Reply:
Has this proven to be a problem for things like [Collectors.mapping(…)](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Collectors.html#mapping(java.util.function.Function,java.util.stream.Collector)) ? 🤔 

There's one implication on turning it into a wildcard—it may actually cause issues implementing the composition directly in the override, as you won't have a typename to refer to.

Do you have a concrete example of where the current "encoding" causes a caller-problem?

Reply:
The API should be client-friendly, not implementor-friendly, given that it's expected to have much more clients than implementors. An implementor can easily delegate to a private method to add missing type parameters. I did exactly this in the [teeing](https://github.com/openjdk/jdk/blob/7d25f1c6cb770e21cfad8096c1637a24e65fab8c/src/java.base/share/classes/java/util/stream/Collectors.java#L1913) Collector before. There should not be a reason to expose a parameter, which is merely an implementation detail. This also complicates the public documentation of the already complex feature, specifying the parameter which is not necessary to specify. 

> Has this proven to be a problem for things like [Collectors.mapping(…)](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/stream/Collectors.html#mapping(java.util.function.Function,java.util.stream.Collector)) ?

I'm not sure how I can prove this, given that you don't usually know the downstream accumulator, so if somebody hits this problem, they will be forced to solve it in another way (e.g., extracting parts of complex collector to separate variables), so you won't find the code in the wild which uses `mapping` type arguments explicitly. Yes, `Collectors.mapping` and `groupingBy` did this mistake in the past, and there's no reason to repeat it again.

> Do you have a concrete example of where the current "encoding" causes a caller-problem?

For example, I want to create a reusable gatherer that performs scan-concat and wraps every resulting string with '[...]'. First try:

```java
var gatherer = Gatherers.scan(() -> "", String::concat)
  .andThen(Gatherer.ofSequential((_, t, pusher) -> pusher.push(STR."[\{t}]")));
```

Now, the type of `gatherer` is `Gatherer<String, ?, Object>`, but I want `Gatherer<String, ?, String>`. I have the following alternatives:
1. Specify the variable type explicitly:
```java
Gatherer<String, ?, String> gatherer = Gatherers.scan(() -> "", String::concat)
    .andThen(Gatherer.ofSequential((_, t, pusher) -> pusher.push(STR."[\{t}]")));
```
2. Specify `ofSequential` type arguments:
```java
var gatherer = Gatherers.scan(() -> "", String::concat)
    .andThen(Gatherer.<String, String>ofSequential((_, t, pusher) -> pusher.push(STR."[\{t}]")));
```
3. Specify the types of lambda arguments explicitly:
```java
var gatherer = Gatherers.scan(() -> "", String::concat)
    .andThen(Gatherer.ofSequential((Void _, String t, Downstream<? super String> pusher) -> pusher.push(STR."[\{t}]")));
```
4. Specify `.andThen` type arguments:
```java
var gatherer = Gatherers.scan(() -> "", String::concat)
    .<Void, String>andThen(Gatherer.ofSequential((_, t, pusher) -> pusher.push(STR."[\{t}]")));
```
In all the cases, I need to specify explicitly more types than I need, as I need to specify only `String`. If we remove `AA`, then specifying `andThen` type arguments would be the shortest and the most readable alternative:

```java
var gatherer = Gatherers.scan(() -> "", String::concat)
    .<String>andThen(Gatherer.ofSequential((_, t, pusher) -> pusher.push(STR."[\{t}]")));
```

Reply:
I agree with Tagir here, it should be an ? and if you want to capture it in a type variable you can use the standard trick of having the private method declaring the type variable and the public method with the wildcard calling the private method.

Reply:
Sold! Type argument removed!